reps/2024-06-13-export-api.md

```diff
@@ -0,0 +1,141 @@
+
+{"event_type": "ACTOR", "actor": {"actor_id": "efc87749b2e337d7872dcf3802000000", "actor_info": {"actorId": "efc87749b2e337d7872dcf3802000000", "jobId": "02000000", "address": {"rayletId": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffff", "workerId": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffff", "ipAddress": "", "port": 0}, "name": "_ray_internal_job_actor_raysubmit_WBMEB9nTaMKmrKrN", "className": "JobSupervisor", "state": "PENDING_CREATION", "numRestarts": "0", "timestamp": 0.0, "pid": 0, "startTime": 0, "endTime": 0, "actorClass": "JobSupervisor", "exitDetail": "-", "requiredResources": {}}}}
+
+```
+
+By aggregating all the Events in time dimension, we can obtain the historical state of the ray cluster,
+and the consumers of the events can aggregate the final state from a series of events. The export api
+does not guarantee the ultimate consistency of event state, which is feasible and tolerated in production practice.
+The loss of some intermediate events will not cause the final state to be incorrect. For example,
+if the event of an actor's creation is lost, but the consumer can read the event of actor exiting,
+then the consumer knows that the actor finally ran successfully.
+
+
+The various components of Ray, including GCS, Dashboard, raylet, and worker, can all serve as event generators.
+However, these sources don't need to concern themselves with the details of how events are stored, transmitted, or published.
+
+### Events to Export and How
+#### Events to Export
+- Core
+Tasks/Actors/Objects/Jobs/Nodes/Placement groups, including its meta data and current status when event emits.
+
+|  Event            | Event Source          | When to export                                                                    | Format example(maybe a file link)|
+|  ----             | ----                  | ----                                                                              | ----                             |
+| Tasks             | CoreWorker and raylet | When coreworker adds task event to<br>ray::core::worker::TaskEventBuffer          |          |
+| Actors            | GCS                   | GcsPublisher::PublishActor<br>publish actor status through<br>GCS_ACTOR_CHANNEL   |          |
+| Objects           | CoreWorker and raylet | None                                                                              |          |
+| Jobs              | JobManager            | When JobInfoStorageClient.put_status called                                       |          |
+| Nodes             | GCS                   | GcsNodeManager::HandleXXXNode                                                     |          |
```

Comment:
Schema for nodes: https://docs.google.com/document/d/1qjoF51h2oUN2sr_MtPnovbNFZYZrh3WLNR_P0HrUuOI/edit?usp=sharing